### PR TITLE
s390x: EVP_CipherInit_ex sequences lead to wrong results

### PR DESCRIPTION
**Pull request by juergenchrist**
The EVP-API allows users to provide parts and even change the direction (decryption vs encryption or vice versa) without providing all arguments.  This led to some errors on s390x where, e.g., a late change in the direction caused the cipher to perform the wrong operation.

This is the master branch equivalent of PR #14900.  For more details on why we want this and where the init sequences come from, please see https://github.com/openssl/openssl/pull/14900#issuecomment-824685249.

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

File: test/recipes/30-test_evp_data/evpciph_aes_common.txt [hunk 1]
```diff
@@ -240,6 +240,7 @@ Operation = ENCRYPT
 Plaintext = 6BC1BEE22E409F96E93D7E117393172A
 Ciphertext = 3B3FD92EB72DAD20333449F8E83CFB4A
 NextIV = 3b3fd92eb72dad20333449f8e83cfb4a
+InvertInit = 1
```
Comment (t8m) [inline test/recipes/30-test_evp_data/evpciph_aes_common.txt:243]: Would it make sense to have the same test cases with and without InvertInit? Or just put the inverted initialization test to one of the evp_extra tests? It does not seem to me to be necessary to have so many test cases where one should trigger the failure already. Or am I wrong?

File: providers/implementations/ciphers/cipher_aes_gcm_hw_s390x.inc [hunk 2]
```diff
@@ -212,6 +224,8 @@ static int s390x_aes_gcm_cipher_update(PROV_GCM_CTX *ctx,
         return 0;
     kma->tpcl = mlen;
 
+    fc = actx->plat.s390x.fc | actx->plat.s390x.hsflag |
```
Comment (slontis) [inline providers/implementations/ciphers/cipher_aes_gcm_hw_s390x.inc:227]: Maybe functionalize this common bit:
````
actx->plat.s390x.fc | actx->plat.s390x.hsflag | S390X_gcm_decrypt_flag(ctx)
````

Comment (juergenchrist) [reply providers/implementations/ciphers/cipher_aes_gcm_hw_s390x.inc]: Done.  Thanks for the suggestion.

File: test/recipes/30-test_evp_data/evpciph_aes_common.txt (hunk 1 above)
Comment (juergenchrist) [reply test/recipes/30-test_evp_data/evpciph_aes_common.txt]: Removed this commit and added some tests to evp_extra_test to test more precisely the errors recently encountered on s390x.  Please have a look at them.

File: test/evp_extra_test.c [hunk 3]
```diff
@@ -2891,6 +2980,388 @@ static int test_names_do_all(void)
...
+    int outlen1, outlen2;
+    int testresult = 0;
+    unsigned char outbuf[1024];
+    unsigned char tag[16];
+    const EVP_INIT_TEST_st *t = &evp_init_tests[idx];
+    EVP_CIPHER_CTX *ctx = NULL;
+    EVP_CIPHER *type = NULL;
+    char *errmsg = NULL;
+
+    ctx = EVP_CIPHER_CTX_new();
```
Comment (slontis) [inline test/evp_extra_test.c:3075]: Tests normally just do

if (TEST_XXX())
   goto err;

File: test/evp_extra_test.c [hunk 4]
```diff
@@ -2891,6 +2980,388 @@ static int test_names_do_all(void)
...
+    }
+    if (!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_GET_TAG, 16, tag)) {
+        errmsg = "GET_TAG1";
+        goto err;
+    }
+    if (!TEST_mem_eq(t->tag1, 16, tag, 16)) {
+        errmsg = "TAG_ERROR1";
+        goto err;
+    }
+    // Now reinit
```
Comment (slontis) [inline test/evp_extra_test.c:3321]: No C++ style comments allowed: Use /* */

File: test/evp_extra_test.c [hunk 5]
```diff
@@ -2891,6 +2980,388 @@ static int test_names_do_all(void)
...
+{
+    int outlen1, outlen2, outlen3;
+    int testresult = 0;
+    unsigned char outbuf[1024];
+    unsigned char tag[16];
+    const TEST_GCM_IV_REINIT_st *t = &gcm_reinit_tests[idx];
+    EVP_CIPHER_CTX *ctx = NULL;
+    EVP_CIPHER *type = NULL;
+    char *errmsg = NULL;
+
```
Comment (slontis) [inline test/evp_extra_test.c:3272]: maybe do
````
size_t taglen = sizeof(tag);
````
And replace the majic number 16 in the code below

File: test/evp_extra_test.c (hunk 3 above)
Comment (t8m) [reply test/evp_extra_test.c:3075]: Yep, although it is rather:
```
if (!TEST_condition(func()))
    goto err;
```
or
```
ret = TEST_condition(func1())
    && TEST_condition(func2())
    .....
```
Please look at the other testcases. These TEST_condition calls really help to diagnose what was the erroneous result.

**Comment by t8m** [comment]
The test failure is relevant.

File: test/evp_extra_test.c (hunk 4 above)
Comment (paulidale) [reply test/evp_extra_test.c:3321]: This is the cause of the test failure.

File: test/evp_extra_test.c (hunk 3 above)
Comment (juergenchrist) [reply test/evp_extra_test.c:3078]: Thanks for pointing this out.  I changed all occurrences but kept the additional info message.

File: test/evp_extra_test.c (hunk 4 above)
Comment (juergenchrist) [reply test/evp_extra_test.c]: Didn't know that.  Sorry.  Fixed.

File: test/evp_extra_test.c (hunk 5 above)
Comment (juergenchrist) [reply test/evp_extra_test.c:3273]: Done.

**Review by t8m — APPROVED** [review]
LGTM

**Comment by juergenchrist** [comment]
One question regarding this:  I also find cipher implementations in crypto/evp/e_aes.c.  These once are still buggy but I could not figure out how to use them with the 3.0 code.  A fix for them for the 1.1.1 branch is in #14900 but I am not sure if it works with 3.0.  Can anybody please advice on how to deal with this file?

**Comment by paulidale** [comment]
Configure and build with `no-asm` and they should be used.

**Review by paulidale — APPROVED** [review]

**Comment by juergenchrist** [comment]
> Configure and build with `no-asm` and they should be used.

I cannot verify that.  I still get the provider implementations used all the time.  Even replacing the `EVP_fetch` in the test with the appropriate `EVP_aes_<keylen>_<mode>()` did not give me the old implementation for long.  The implicit fetch in `evp_cipher_init_internal` gives me again the new version.  From the code I think the only way to get the old code is if an engine gives exactly this cipher since in that case we skip the whole provider code.

If that is true I have no idea how I can use the old versions.  So the old code looks pretty dead to me and I do not think we need to care about it.  But anyway, the fixes in #14900 could be used here, too.

**Comment by slontis** [comment]
> From the code I think the only way to get the old code is if an engine gives exactly this cipher since in that case we skip the whole provider code.

You are correct. Engines are the only reason most of the legacy code is still needed.

**Comment by paulidale** [comment]
Merged to master, thanks for the fix.

**Comment by p-steuer** [comment]
> You are correct. Engines are the only reason most of the legacy code is still needed.

If an engine implements eg aes-256-gcm how would the legacy platform specific code in e_aes.c be used ?

If the platformspecific code in e_aes.c is still somehow relevant in master, first commit from https://github.com/openssl/openssl/pull/14900 would be needed for master as well.

**Pull request closed without merging**